addons/website_sale_autocomplete/controllers/main.py

```diff
@@ -0,0 +1,151 @@
+# -*- encoding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import requests
+
+from odoo import http
+from odoo.http import request
+from odoo.tools import html2plaintext
+
+FIELDS_MAPPING = {
+    'country': ['country'],
+    'street_number': ['number'],
+    'locality': ['city'],  # If locality exists, use it instead of the more general administrative area
+    'route': ['street'],
+    'postal_code': ['zip'],
+    'administrative_area_level_1': ['state', 'city'],
+    'administrative_area_level_2': ['state', 'country']
+}
+
+# If a google fields may correspond to multiple standard fields, the first occurrence in the list will overwrite following entries.
+FIELDS_PRIORITY = ['country', 'street_number', 'neighborhood', 'locality', 'route', 'postal_code',
+                   'administrative_area_level_1', 'administrative_area_level_2']
+GOOGLE_PLACES_ENDPOINT = 'https://maps.googleapis.com/maps/api/place'
+TIMEOUT = 2.5
+
+
+class AutoCompleteController(http.Controller):
+
+    def _get_key(self):
+        return request.env['website'].get_current_website().google_places_api_key
+
+    def _translate_google_to_standard(self, google_fields):
+        standard_data = {}
+
+        for google_field in google_fields:
+            fields_standard = FIELDS_MAPPING[google_field['type']] if google_field['type'] in FIELDS_MAPPING else []
+
+            for field_standard in fields_standard:
+                if field_standard in standard_data:  # if a value is already assigned, do not overwrite it.
+                    continue
+                # Convert state and countries to odoo ids
+                if field_standard == 'country':
+                    standard_data[field_standard] = request.env['res.country'].search(
+                        [('code', '=', google_field['short_name'].upper())])[0].id
+                elif field_standard == 'state':
+                    state = request.env['res.country.state'].search(
+                        [('code', '=', google_field['short_name'].upper()),
+                         ('country_id.id', '=', standard_data['country'])])
+                    if len(state) == 1:
+                        standard_data[field_standard] = state.id
+                else:
+                    standard_data[field_standard] = google_field['long_name']
+        return standard_data
+
+    def _guess_number_from_input(self, source_input, standard_address):
+        """
+        Google might not send the house number in case the address
+        does not exist in their database.
+        We try to guess the number from the user's input to avoid losing the info.
+        """
+        # Remove other parts from address to make better guesses
+        guess = source_input.replace(standard_address.get('zip', ''), '') \
```

Comment:
```suggestion
        guess = source_input \
            .replace(standard_address.get('zip', ''), '') \
```
should be uniform for clarity / readability.